src/aws_credentials_ec2.erl

```diff
@@ -3,40 +3,70 @@
 -module(aws_credentials_ec2).
 -behaviour(aws_credentials_provider).
 
+-define(SESSION_TOKEN_URL,
+        "http://169.254.169.254/latest/api/token/").
+-define(SESSION_TOKEN_TTL_HEADER,
+        "x-aws-ec2-metadata-token-ttl-seconds").
+-define(SESSION_TOKEN_TTL_SECONDS,
+        "21600").
+-define(SESSION_TOKEN_HEADER,
+        "X-aws-ec2-metadata-token").
 -define(CREDENTIAL_URL,
         "http://169.254.169.254/latest/meta-data/iam/security-credentials/").
 -define(DOCUMENT_URL,
         "http://169.254.169.254/latest/dynamic/instance-identity/document").
 
 -type role() :: binary().
+-type session_token() :: binary().
+% TODO: How can I get these types from httpc?
+-type field() :: string().
+-type value() :: string().
+-type request_header() :: {field(), value()}.
+-type request_headers() :: [request_header()].
 
 -export([fetch/1]).
 
 -spec fetch(aws_credentials_provider:options()) ->
         {ok, aws_credentials_provider:credentials(), aws_credentials_provider:expiration()} |
         {error, any()}.
 fetch(_Options) ->
-    {ok, Role} = fetch_role(),
-    {ok, AccessKeyID, SecretAccessKey, ExpirationTime, Token} = fetch_metadata(Role),
-    {ok, Region} = fetch_document(),
+    % Fetch the IMDS session token and convert it to the request headers.
+    % Then pass the request headers to subsequent IMDS requests.
+    {ok, SessionToken} = fetch_session_token(),
+    {ok, RequestHeaders} = request_headers(SessionToken),
+    {ok, Role} = fetch_role(RequestHeaders),
+    {ok, AccessKeyID, SecretAccessKey, ExpirationTime, Token} = fetch_metadata(Role, RequestHeaders),
+    {ok, Region} = fetch_document(RequestHeaders),
     Credentials = aws_credentials:make_map(?MODULE, AccessKeyID, SecretAccessKey, Token, Region),
     {ok, Credentials, ExpirationTime}.
 
--spec fetch_role() -> {ok, role()}.
-fetch_role() ->
-    {ok, 200, Body, _Headers} = aws_credentials_httpc:get(?CREDENTIAL_URL),
+-spec fetch_session_token() -> {ok, session_token()}.
+fetch_session_token() ->
+  RequestHeaders = [{?SESSION_TOKEN_TTL_HEADER, ?SESSION_TOKEN_TTL_SECONDS}],
+  {ok, 200, Body, _Headers} = aws_credentials_httpc:put(?SESSION_TOKEN_URL, RequestHeaders),
     {ok, Body}.
 
-fetch_metadata(Role) ->
-    Url = lists:flatten(io_lib:format("~p/~p", [?CREDENTIAL_URL, Role])),
```

Comment:
This part right here is what was broken on master, from what I could tell.

Reply:
Indeed, it should be fixed now.